src/realm/sync/subscriptions.hpp

```diff
@@ -148,12 +150,12 @@ class SubscriptionSet {
         iterator& operator++();
         iterator operator++(int);
 
-    protected:
+    private:
         friend class SubscriptionSet;
+        friend class MutableSubscriptionSet;
```

Comment:
My preference is to try to reduce the number of friends. Having inheritance with bidirectional friendship seems like a bit of a smell. If it is not enough to let the const methods do their job and you need to hide the mutation methods from the public API, could you move `MutableSubscriptionSet` entirely into the cpp file? If so, I think you could remove `make_mutable_copy()` and have `Subscription` directly call the `MutableSubscriptionSet` constructor where needed. This should remove the need for bidirectional friendship between `SubscriptionSet` and `MutableSubscriptionSet`. What do you think?

Reply:
I'm all for fewer friends! The purpose of this PR isn't quite to make MutableSubscriptionSet hidden from the public API tho - in the old model we return a `const SubscriptionSet` to hide the mutable functions, but it was pointed out to me that if you do `auto sub_set = store->get_latest();` the type of `sub_set` is just `SubscriptionSet`.  So this PR splits the type in two so that it's clear based on the type of object what you're allowed to do with it. I've taken another crack at this and hopefully it's clearer/less friendly in a good way.